x-pack/plugins/canvas/server/lib/route_expression/thread/index.js

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { fork } from 'child_process';
+import { resolve } from 'path';
+import uuid from 'uuid/v4';
+import { serializeProvider } from '../../../../common/lib/serialize';
+import { populateServerRegistries } from '../../populate_server_registries';
+
+const serialization = populateServerRegistries(['types']).then(({ types }) =>
+  serializeProvider(types.toJS())
+);
+
+// If the worker doesn't response in 10s, kill it.
+const WORKER_TIMEOUT = 10000;
+const workerPath = resolve(__dirname, 'babeled.js');
+const heap = {};
+let worker = null;
+
+function getWorker() {
+  if (worker) return worker;
+  worker = fork(workerPath, {});
+
+  // 'exit' happens whether we kill the worker or it just dies.
+  // No need to look for 'error', our worker is intended to be long lived so it isn't running, it's an issue
+  worker.on('exit', () => {
+    worker = null;
+    worker = getWorker();
+  });
+
+  worker.on('message', msg => {
+    const { type, value, id } = msg;
+    if (type === 'run') {
+      const { threadId } = msg;
+      const { ast, context } = value;
+      heap[threadId].onFunctionNotFound(ast, context).then(value => {
```

Comment:
I don't think we need onFunctionNot found on the heap, there's nothing special about this onFunctionNotFound. If this is true then we also don't need to track `threadId`?